### PR TITLE
remove duplicated queries

### PR DESCRIPTION
**Pull request by alikon**
### Summary of Changes
the same query is executed once

### Testing Instructions
- in a multilanguage site
- enable debug
- https://github.com/joomla/joomla-cms/pull/19682#issuecomment-385151852
- look at debug database duplicated queries

### Expected result
no duplicated queries

### Actual result
duplicate queries one of them is this one

screenshot from 2018-02-14 21-04-06

### CODE REVIEW

File: administrator/components/com_categories/helpers/categories.php [hunk 1]
```diff
@@ -110,34 +118,42 @@ public static function getActions($extension, $categoryId = 0)
...
 			$db    = \JFactory::getDbo();
 
 			$query = $db->getQuery(true)
-				->select($db->qn('published'))
+				->select($db->qn(array('id', 'language') ))
 				->from($db->qn('#__categories'))
-				->where('access IN (' . $groups . ')')
-				->where($db->qn('id') . ' = ' . (int) $assocId);
+				->where($db->qn('id') . ' IN (' . $catId . ')')
+				->where($db->qn('published') . ' =  1')
```
Comment (Quy) [inline administrator/components/com_categories/helpers/categories.php]: Remove extra space after `=`

File: administrator/components/com_categories/helpers/categories.php [hunk 2]
```diff
@@ -110,34 +118,42 @@ public static function getActions($extension, $categoryId = 0)
...
+			$groups           = implode(',', $user->getAuthorisedViewLevels());
+
+			foreach ($langAssociations as $langAssociation)
+			{
+				// Include only published categories with user access
+				$arrId    = explode(':', $langAssociation->id);
+				$assocId  = $arrId[0];
+			}
+
+			$catId    = implode(',', $assocId);
```
Comment (Quy) [inline administrator/components/com_categories/helpers/categories.php]: Remove extra spaces before `=`

File: administrator/components/com_categories/helpers/categories.php (hunk 1 above)
Comment (alikon) [reply administrator/components/com_categories/helpers/categories.php]: done

File: administrator/components/com_categories/helpers/categories.php (hunk 2 above)
Comment (alikon) [reply administrator/components/com_categories/helpers/categories.php]: done

**Comment by ethernidee** [comment]
What do you think about renaming `$filters` to something like `$cachedAssociations`? Perhaps, it will be clearer, what is stored inside.

    FILE: ...omla-cms/administrator/components/com_categories/helpers/categories.php
    --------------------------------------------------------------------------------
    FOUND 1 ERROR(S) AFFECTING 1 LINE(S)
    --------------------------------------------------------------------------------
    140 | ERROR | Space before closing parenthesis of function call prohibited

**Comment by alikon** [comment]
from my personal pov is explained in declaration `Cached array of the category item id.`
but if you feel that renaming is better .... the pr branch is public....

File: administrator/components/com_categories/helpers/categories.php [hunk 3]
```diff
@@ -110,34 +118,42 @@ public static function getActions($extension, $categoryId = 0)
...
+			$langAssociations = JLanguageAssociations::getAssociations($extension, '#__categories', 'com_categories.item', $pk, 'id', 'alias', '');
+			$associations     = array();
+			$user             = JFactory::getUser();
+			$groups           = implode(',', $user->getAuthorisedViewLevels());
+
+			foreach ($langAssociations as $langAssociation)
+			{
+				// Include only published categories with user access
+				$arrId    = explode(':', $langAssociation->id);
+				$assocId  = $arrId[0];
```
Comment (Quy) [inline administrator/components/com_categories/helpers/categories.php]: Isn’t  `$assocId` being reassigned each time and not array as to be expected?

Comment (alikon) [reply administrator/components/com_categories/helpers/categories.php]: right, fixed

File: administrator/components/com_categories/helpers/categories.php [hunk 4]
```diff
@@ -110,34 +118,42 @@ public static function getActions($extension, $categoryId = 0)
...
+			$langAssociations = JLanguageAssociations::getAssociations($extension, '#__categories', 'com_categories.item', $pk, 'id', 'alias', '');
+			$associations     = array();
+			$user             = JFactory::getUser();
+			$groups           = implode(',', $user->getAuthorisedViewLevels());
+
+			foreach ($langAssociations as $langAssociation)
+			{
+				// Include only published categories with user access
+				$arrId      = explode(':', $langAssociation->id);
+				$assocId[]  = $arrId[0];
```
Comment (Quy) [inline administrator/components/com_categories/helpers/categories.php]: Is it necessary to initialize `$assocId` outside the loop?

Remove extra space before `=`.

**Comment by Quy** [comment]
The return values are different.

Before PR:
```
array(2) {
  ["fr-FR"]=>
  string(18) "11:categorie-fr-fr"
  ["es-ES"]=>
  string(18) "12:categoria-es-es"
}
```
After PR:
```
array(2) {
  ["fr-FR"]=>
  string(2) "11"
  ["es-ES"]=>
  string(2) "12"
}
```

**Comment by ladyjer** [comment]
I have tested this item :white_check_mark: successfully on 7d724e02d8635d8e03965cdcb9d0f8a707bb4c94<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/19682">issues.joomla.org/tracker/joomla-cms/19682</a>.</sub>

**Comment by Quy** [comment]
I have tested this item :white_check_mark: successfully on a74c9a3aea31eeab9ba267b6f2882978c7818bea<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/19682">issues.joomla.org/tracker/joomla-cms/19682</a>.</sub>

**Comment by ghost** [comment]
@ladyjer can you please retest?

**Comment by infograf768** [comment]
Notice: Undefined variable: assocId in /Applications/MAMP/htdocs/installmulti/trunkgitnew/administrator/components/com_categories/helpers/categories.php on line 135

Warning: implode(): Invalid arguments passed in /Applications/MAMP/htdocs/installmulti/trunkgitnew/administrator/components/com_categories/helpers/categories.php on line 135

1064 You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near ') AND `published` = 1 AND `access` IN (1,1,5)' at line 3

File: administrator/components/com_categories/helpers/categories.php [hunk 5]
```diff
@@ -110,34 +118,48 @@ public static function getActions($extension, $categoryId = 0)
...
 		{
-			// Include only published categories with user access
-			$arrId    = explode(':', $langAssociation->id);
-			$assocId  = $arrId[0];
+			$langAssociations = JLanguageAssociations::getAssociations($extension, '#__categories', 'com_categories.item', $pk, 'id', 'alias', '');
+			$associations     = array();
+			$user             = JFactory::getUser();
+			$groups           = implode(',', $user->getAuthorisedViewLevels());
+
+			// No assocations
```
Comment (infograf768) [inline administrator/components/com_categories/helpers/categories.php]: No associations

**Comment by infograf768** [comment]
Test instructions should be updated.
This PR (after last correction) is checking duplicate queries of associated categories, NOT category menu items.
We can indeed associate category menu items without ever associating the categories concerned.

Therefore, to test, one should not associate the menu items, only the categories concerned.

Needs totally new tests.

**Comment by alikon** [comment]
updated 

> Needs totally new tests.

:+1:

**Comment by Quy** [comment]
I have tested this item :white_check_mark: successfully on 2d303622f40443181690d44bca10aabe9f9d0ff8<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/19682">issues.joomla.org/tracker/joomla-cms/19682</a>.</sub>

**Review by infograf768 — APPROVED** [review]

**Comment by csthomas** [comment]
@alikon Are you sure we need `static::$filters` at all?

Before, the sql query was executed inside a loop, now it is executed outside. I have not yet tested it.

**Comment by viocassel** [comment]
I have tested this item :white_check_mark: successfully on 2d303622f40443181690d44bca10aabe9f9d0ff8<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/19682">issues.joomla.org/tracker/joomla-cms/19682</a>.</sub>

**Comment by Quy** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/19682">issues.joomla.org/tracker/joomla-cms/19682</a>.</sub>

**Comment by alikon** [comment]
@HLeithner please a final response 

1 year old rtc pr

**Comment by HLeithner** [comment]
@alikon thx for you work but this is an optimization that should go into j4 and not in j3 because it doesn't fix a bug...

If this query still exists in j4 it would be great if you could rebase this pr to 4.0-dev

ymmv

**Pull request closed without merging**